crates/testing/src/autodiscover.rs

```diff
@@ -223,3 +232,198 @@ fn run_pytest(crate_dir: &Path) -> Result<(), Error> {
+
+    let jest_file_name = "jest.config.js";
+    let jest_config_file = crate_dir.join(jest_file_name);
+
+    let config_content = r#"/** @type {import('jest').Config} */
+const config = {
+    verbose: true,
+    testPathIgnorePatterns: ["node_modules", "snapshots"],
+};
+module.exports = config;"#;
+
+    fs::write(&jest_config_file, config_content)?;
+    anyhow::ensure!(crate_dir.join(&jest_config_file).exists());
+
+    // reading the package and getting the namespace and name of the javascript created package
+    let package_path = generated_bindings.join("package");
+    tracing::info!(?package_path, "package_path for generated hello-wasi");
+
+    let mut cmd = Command::new("yarn");
+    cmd.current_dir(&package_path);
+
+    tracing::info!(?cmd, "Installing dependencies for generated bindings");
+    let status = cmd
+        .stdin(Stdio::null())
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .current_dir(&package_path)
+        .status()
+        .context("Unable to run yarn. Is it installed?")?;
+    anyhow::ensure!(
+        status.success(),
+        "Unable to install dependencies for generated bindings"
+    );
+
+    let package_name = get_package_name(&package_path)?;
+
+    let mut cmd = Command::new("yarn");
+    cmd.arg("link").current_dir(&package_path);
+
+    tracing::info!(?cmd, "Linking the generated bindings as a `Yarn link`");
+    let status = cmd
+        .stdin(Stdio::null())
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .current_dir(&package_path)
+        .status()
+        .context("Unable to run yarn. Is it installed?")?;
+    anyhow::ensure!(
+        status.success(),
+        "Unable to perform yarn link on generated bindings"
+    );
+
+    let mut cmd = Command::new("yarn");
+    cmd.arg("link").arg(&package_name).current_dir(crate_dir);
+
+    tracing::info!(?cmd, "Linking the testing package to generated bindings");
+    let status = cmd
+        .stdin(Stdio::null())
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .current_dir(crate_dir)
+        .status()
+        .context("Unable to run yarn. Is it installed?")?;
+    anyhow::ensure!(
+        status.success(),
+        "Unable to initialize a link to the generated bindings from testing crate"
+    );
+    Ok(())
+}
+
+fn run_jest(crate_dir: &Path, generated_bindings: &Path) -> Result<(), Error> {
+    let mut cmd = Command::new("yarn");
+    cmd.arg("jest").current_dir(crate_dir);
+
+    tracing::info!(?cmd, "Running Jest tests");
+    let status = cmd
+        .stdin(Stdio::null())
+        .stdout(Stdio::inherit())
+        .stderr(Stdio::inherit())
+        .current_dir(crate_dir)
+        .status()
+        .context("Unable to run yarn. Is it installed?")?;
+    anyhow::ensure!(status.success(), "jest failed");
+
+    let _ = YarnConfig {
+        crate_dir,
+        generated_bindings,
+    };
```

Comment:
I don't think this is what you wanted to do because assigning something to `_` will cause it to be immediately dropped and call the destructor.

What you probably wanted to do is return the `YarnConfig` from `setup_javascript()` and save it to a variable (e.g. `let _guard = setup__javascript(...)?`) inside the ` Language::JavaScript => { ... }` match arm. That way our `YarnConfig` will be created while setting up and only get dropped after running `run_jest()` has finished.

As it is, any errors that occur between the initial `yarn link` in `setup_javascript()` and the final `anyhow::ensure!(status.success(), "jest failed")` in `run_jest()` will mean we never run `yarn unlink` (because the `YarnConfig` isn't created until just before we return from `run_jest()`).

Reply:
got it. I'll keep that in mind, currently removing the yarn unlink thing altogether.